File: api/src/main/java/org/pmiops/workbench/cohortbuilder/util/Validation.java
package org.pmiops.workbench.cohortbuilder.util;

import java.text.MessageFormat;
import java.util.function.Predicate;
import org.pmiops.workbench.exceptions.BadRequestException;

public class Validation<K> {

  private Predicate<K> predicate;
  private Boolean throwException;

  public static <K> Validation<K> from(Predicate<K> predicate) {
    return new Validation<K>(predicate);
  }

  private Validation(Predicate<K> predicate) {
    this.predicate = predicate;
  }

  public Validation test(K param) {
    return predicate.test(param) ? throwException() : ok();
  }

  public Validation throwException() {
    this.throwException = true;
    return this;
  }

  public Validation ok() {
    this.throwException = false;
    return this;
  }

  public void throwException(String message) {
    if (throwException) {
      throw new BadRequestException(message);
    }
  }

  public void throwException(String message, Object... args) {
    if (throwException) {
      throw new BadRequestException(new MessageFormat(message).format(args));
    }
  }
}

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/util/Validation.kt
package org.pmiops.workbench.cohortbuilder.util

import java.text.MessageFormat
import java.util.function.Predicate
import org.pmiops.workbench.exceptions.BadRequestException

class Validation<K> private constructor(private val predicate: Predicate<K>) {
    private var throwException: Boolean? = null

    fun test(param: K): Validation<*> {
        return if (predicate.test(param)) throwException() else ok()
    }

    fun throwException(): Validation<*> {
        this.throwException = true
        return this
    }

    fun ok(): Validation<*> {
        this.throwException = false
        return this
    }

    fun throwException(message: String) {
        if (throwException!!) {
            throw BadRequestException(message)
        }
    }

    fun throwException(message: String, vararg args: Any) {
        if (throwException!!) {
            throw BadRequestException(MessageFormat(message).format(args))
        }
    }

    companion object {

        fun <K> from(predicate: Predicate<K>): Validation<K> {
            return Validation(predicate)
        }
    }
}
